### PR TITLE
Fix: Infinite Loader Threshold

### PR DESCRIPTION
**Pull request by dmitriy-kudelko**
Improves the mechaninsm of calculating the threshold to be based on row index rather than on scroll percentage which could lead to false positives

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -121,22 +121,22 @@ Sticky React Table supports a host of properties which allow you to completely c
...
+| selectedRows                 | `Array`              | `undefined` | An array of ids to make the table act as a controlled selection component.                          | No       |
+| infiniteScrollTotalCount     | `Number`             | `undefined` | Total number of rows which can be loaded using infinite loader.                                     | No       |
+| infiniteScrollLoadMore       | `Function`           | `undefined` | Invoked when a new page is requested when using infinite loader.                                    | No       |
+| infiniteScrollThreshold      | `Number`             | `10`        | The number of rows before the end of the table that will trigger a call to infiniteScrollThreshold. | No       |
+| infiniteScrollLoaderRowCount | `Number`             | `5`         | The number of additional rows to display at the bottom for progress indication.                     | No       |
+| infiniteScrollPageSize       | `Number`             | `30`        | A number of rows loaded with each portion.                                                          | No       |
```
Comment (samrith-s) [inline README.md:138]: Is this an internal prop? What is the use of this and why is this required for the table? Afaik, table shouldn't really need to know how much we want to load. Just take in the new data, and render it. If the keys provided are correct, it won't rerender all the rows.

File: src/components/Table.js [hunk 2]
```diff
@@ -391,43 +400,54 @@ export default class Table extends PureComponent {
...
-        const nextPage = data.length / infiniteScrollPageSize + 1;
+    if (!this.requestedPages[nextPage]) {
+      const targetRow = this.rowRefs[dataCount - infiniteScrollThreshold];
 
-        if (!this.requestedPages[nextPage]) {
-          this.requestedPages[nextPage] = true;
+      const {
+        top: innerTop,
+        height: viewPortHeight
+      } = this.innerRef.getBoundingClientRect();
```
Comment (samrith-s) [inline src/components/Table.js:423]: Maybe we should give access to the `innerRef` so that user can scroll to top and do some stuff like that.

**Review by samrith-s — CHANGES_REQUESTED** [review]
@dmitriy-kudelko Just a couple of questions below. Once that's resolved, can merge it. Looks good otherwise.

File: README.md (hunk 1 above)
Comment (dmitriy-kudelko) [reply README.md:138]: No it's not an internal one. it's needed to determine the next page number to load which in turn is passed to `infiniteScrollLoadMore` call.  It's also used internally to track which pages were already requested. Otherwise we'll have to calculate the offset in each component the Table is used. In fact you never load some arbitrary row count (except for the last portion) - you exactly know how may rows are returned by the BE.

File: src/components/Table.js (hunk 2 above)
Comment (dmitriy-kudelko) [reply src/components/Table.js:423]: Yes. It's already accessible as `<Table ref={ref => ref.innerRef} />` cause it's an instance prop.

**Comment by dmitriy-kudelko** [comment]
@samrith-s , please check replies.

File: README.md (hunk 1 above)
Comment (samrith-s) [reply README.md:138]: I think in this, we might be doing a lot more than what we should, within the scope of the table. The fact that we natively support infinite loading is a feature enough. I think all of this should be handled by the user, and not by us. 

I genuinely believe we are overcomplicating a rather straightforward feature. 

What are your thoughts regarding the same?

Comment (dmitriy-kudelko) [reply README.md:138]: I think it's still a very basic implementation. This is how many infinite loaders work taking either page number or number of rows to load (which is in some sense the same stuff) and doing calculations based on it. 

Again we need to track current page inside the component for a few reasons:
- to detect that a certain page was requested - to avoid multiple requests of the same page
- and what's really important for scroll performance - to do less calculations in `onScroll` handler: https://github.com/samrith-s/sticky-react-table/blob/a2b4dab62667b211fa7f93408ebc1cfdb3e4f553/src/components/Table.js#L417

I believe that doing all this outside is gonna be error prone and repetitive. 

Take a look and react-virtualized `InfiniteLoader.minimumBatchSize` prop - it's the same stuff.

Comment (samrith-s) [reply README.md:138]: Hey,

I was looking on it before your comment. What we can do is refactor this a bit to reduce the props or rename them a bit better in future versions. For now, let's go with this.

**Review by samrith-s — APPROVED** [review]
Looks good. Will make a new issue for refactoring the props, and maybe the working a little bit.

File: README.md (hunk 1 above)
Comment (dmitriy-kudelko) [reply README.md:138]: @samrith-s  I honestly don't think this is the final implementation ) So we'll have a bunch of time to think how to improve it depending on use cases :)
